predicators/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -39,12 +40,159 @@ def _create_general_pnad_for_option(
-        preconditions = self._induce_preconditions_via_intersection(pnad)
-        pnad.op = pnad.op.copy_with(preconditions=preconditions)
-
         return pnad
 
+    def spawn_new_pnad(self, segment: Segment) -> PartialNSRTAndDatastore:
+        """Given some segment with necessary add effects that a new PNAD must
+        achieve, create such a PNAD ("spawn" from the most general one
+        associated with the segment's option) so that it has the necessary add
+        effects contained in the given segment."""
+        # Create a general PNAD for the segment's option.
+        pnad = self._create_general_pnad_for_option(
+            segment.get_option().parent)
+        # Assert that this really is a general PNAD.
+        assert len(pnad.op.add_effects) == 0, \
+            "Can't spawn from non-general PNAD"
+        # Assert that the segment contains necessary_add_effects.
+        necessary_add_effects = segment.necessary_add_effects
+        assert necessary_add_effects is not None
+
+        # Get an arbitrary grounding of the PNAD's operator whose
+        # preconditions hold in segment.init_atoms.
+        objects = set(segment.states[0])
+        _, var_to_obj = self._find_best_matching_pnad_and_sub(
+            segment, objects, [pnad], check_only_preconditions=True)
+        # Assert that such a grounding exists - this must be the case
+        # since we only ever call this method with the most general
+        # PNAD for the option.
+        assert var_to_obj is not None
+        obj_to_var = {v: k for k, v in var_to_obj.items()}
+        assert len(var_to_obj) == len(obj_to_var)
+        # Before we can lift the necessary_add_effects, we need to add new
+        # entries to obj_to_var, since necessary_add_effects may
+        # contain objects that were not in the ground operator's
+        # parameters.
+        all_objs = {o for eff in necessary_add_effects for o in eff.objects}
+        missing_objs = sorted(all_objs - set(obj_to_var))
+        new_vars = utils.create_new_variables([o.type for o in missing_objs],
+                                              existing_vars=pnad.op.parameters)
+        obj_to_var.update(dict(zip(missing_objs, new_vars)))
+        # Finally, we can lift necessary_add_effects.
+        updated_params = sorted(obj_to_var.values())
+        updated_add_effects = {
+            a.lift(obj_to_var)
+            for a in necessary_add_effects
+        }
+
+        # Create a new PNAD with the given parameters and add effects. Set
+        # the preconditions to be trivial. They will be recomputed later.
+        new_pnad_op = pnad.op.copy_with(parameters=updated_params,
+                                        preconditions=set(),
+                                        add_effects=updated_add_effects)
+        new_pnad = PartialNSRTAndDatastore(new_pnad_op, [], pnad.option_spec)
+        # Note: we don't need to copy anything related to keep effects into
```

Comment:
update this comment? there isn't such thing as a "final version of backchaining" in the new subclass right?